src/util.rs

```diff
@@ -11,9 +11,20 @@ pub fn cp_r(src: &Path, dst: &Path) -> Result<()> {
     (|| -> io::Result<()> {
             for e in src.read_dir()? {
                 let e = e?;
+                let metadata = e.path().metadata()?;
                 let f = e.file_name();
+                let src = src.join(&f);
+                let dst = dst.join(&f);
 
-                fs::copy(src.join(&f), dst.join(&f))?;
+                if metadata.is_dir() {
+                    fs::create_dir(&dst)?;
+                    cp_r(
+                        src.as_path(),
+                        dst.as_path()
+                    ).map_err(|e| io::Error::new(io::ErrorKind::Other, e.description()))?;
```

Comment:
This is an interesting situtation! I think to avoid the back and forth between different error types you could write an "inner" function that does all the logic but always returns io::Result and then have cp_r call that inner function with chain_err. Something like this (not tested):

``` rust
pub fn cp_r(src: &Path, dst: &Path) -> Result<()> {
    fn cp_r_(src: &Path, dst: &Path) -> io::Result<()> {
        for e in src.read_dir()? {
            let e = e?;
            let metadata = e.path().metadata()?;
            let f = e.file_name();
            let src = src.join(&f);
            let dst = dst.join(&f);

            if metadata.is_dir() {
                fs::create_dir(&dst)?;
                cp_r_(src.as_path(), dst.as_path())?;
            } else {
                fs::copy(src, dst)?;
            }
        }

        Ok(())
    }

    cp_r_(src, dst).chain_err(
        || {
            format!(
                "copying files from `{}` to `{}` failed",
                src.display(),
                dst.display()
            )
        },
    )
}
```